compiler/optimizer/LocalOpts.cpp

```diff
@@ -3548,6 +3548,27 @@ int32_t TR_EliminateRedundantGotos::process(TR::TreeTop *startTree, TR::TreeTop
             !performTransformation(comp(), "%sEliminating goto at the end of block_%d with BBStart %p\n", optDetailString(), block->getNumber(), block->getEntry()->getNode()))
         continue;
 
+      // Blocks could be removed when the exception edges are removed from the block
+      // The IN edges saved in fixablePreds need to be updated
+      for (auto previousEdge = fixablePreds.before_begin(), currentEdge = fixablePreds.begin();
+           currentEdge != fixablePreds.end();
+           ++previousEdge, ++currentEdge)
+         {
+         TR::CFGEdge* current = *currentEdge;
+         TR::Block *from = toBlock(current->getFrom());
+         TR::Block *to = toBlock(current->getTo());
+
+         if (!from->isValid() || !to->isValid())
```

Comment:
It shouldn't be possible for `to` to have become unreachable due to the removal of exception edges outgoing from `block` above. The `current` edge comes from `fixablePreds`, so it must have been found originally in `block->getPredecessors()`, and therefore `to == block`. Because `block` was in the CFG before removing the exception edges, it must have been reachable, i.e. there was a path from the entry to `block`. That path does not contain any edges outgoing from `block`, and in particular it doesn't contain any of the removed exception edges, because if it did, then `block` would appear multiple times in the path. So once the exception edges are removed, that path survives, and `block` is still reachable in the modified CFG

This is good because we'd probably have a bigger problem if `block` could disappear on us!

I suggest ignoring `to->isValid()` in the condition here, and maybe adding some assertions to verify e.g. `to == block` and `block->isValid()`. This would also make it unnecessary to log the value of `to->isValid()`